File: problems/1802_maximum_value_at_a_given_index_in_a_bounded_array/solution.py
class Solution:
    def maxValue(self, n: int, index: int, maxSum: int) -> int:
        def count(mid: int) -> int:
            left_count = min(mid - 1, index)
            right_count = min(mid - 1, n - 1 - index)
            total = mid
            if left_count < index:
                total += left_count * (left_count + 1) // 2 + (index - left_count)
            else:
                total += left_count * (left_count + 1) // 2 - (left_count - index) * (left_count - index + 1) // 2
            if right_count < n - 1 - index:
                total += right_count * (right_count + 1) // 2 + (n - 1 - index - right_count)
            else:
                total += right_count * (right_count + 1) // 2 - (right_count - (n - 1 - index)) * (right_count - (n - 1 - index) + 1) // 2
            return total
        lo, hi = 1, maxSum
        while lo < hi:
            mid = (lo + hi + 1) // 2
            if count(mid) <= maxSum:
                lo = mid
            else:
                hi = mid - 1
        return lo

File: problems/1802_maximum_value_at_a_given_index_in_a_bounded_array/Solution.java
class Solution {
    public int maxValue(int n, int index, int maxSum) {
        int lo = 1, hi = maxSum;
        while (lo < hi) {
            int mid = lo + (hi - lo + 1) / 2;
            if (count(n, index, mid) <= maxSum) lo = mid;
            else hi = mid - 1;
        }
        return lo;
    }
    private long count(int n, int index, int mid) {
        long total = mid;
        long left = Math.min(mid - 1, index);
        long right = Math.min(mid - 1, n - 1 - index);
        total += left * (left + 1) / 2;
        if (left < index) total += index - left;
        else total -= (left - index) * (left - index + 1) / 2;
        total += right * (right + 1) / 2;
        if (right < n - 1 - index) total += n - 1 - index - right;
        else total -= (right - (n - 1 - index)) * (right - (n - 1 - index) + 1) / 2;
        return total;
    }
}
